### PR TITLE
Handle non ASCII headers

### PR DESCRIPTION
**Pull request by leonardo-piseri**
See closed pull request #16 
The new code should now pass tests.

### CODE REVIEW

File: src/Headers.php [hunk 1]
```diff
@@ -152,7 +152,7 @@ public function setEncoding($encoding)
     {
         $this->encoding = $encoding;
         foreach ($this as $header) {
-            $header->setEncoding($encoding);
+            ($encoding === 'ASCII') ? $header->getEncoding() : $header->setEncoding($encoding);
```
Comment (Maks3w) [inline src/Headers.php]: why call `getEncoding`? breaks the lazy load idea

File: test/HeadersTest.php [hunk 2]
```diff
@@ -424,4 +424,22 @@ public function testHeaderCrLfAttackAddHeadersMultiple()
         $headers->addHeaders(['Fake' => ["foo-bar\r\n\r\nevilContent"]]);
         $headers->forceLoading();
     }
+    
+    public function testCanConvertNonASCIIHeaderToStringAfterAddingHeaderToMessage()
+    {
+        //non ASCII characters
+        $str = 'òàù';
+        $header = new \Zend\Mail\Header\GenericHeader();
+        $message = new \Zend\Mail\Message();
```
Comment (Maks3w) [inline test/HeadersTest.php]: What is the purpose of this object in the test?

File: test/HeadersTest.php [hunk 3]
```diff
@@ -424,4 +424,22 @@ public function testHeaderCrLfAttackAddHeadersMultiple()
         $headers->addHeaders(['Fake' => ["foo-bar\r\n\r\nevilContent"]]);
         $headers->forceLoading();
     }
+    
+    public function testCanConvertNonASCIIHeaderToStringAfterAddingHeaderToMessage()
+    {
+        //non ASCII characters
+        $str = 'òàù';
+        $header = new \Zend\Mail\Header\GenericHeader();
```
Comment (Maks3w) [inline test/HeadersTest.php]: Zend\Mail namespace is already imported. Please use it.

File: test/HeadersTest.php [hunk 4]
```diff
@@ -424,4 +424,22 @@ public function testHeaderCrLfAttackAddHeadersMultiple()
...
+        $message = new \Zend\Mail\Message();
+        $headers = new \Zend\Mail\Headers();
+        
+        $header->setFieldName('X-Test');
+        $header->setFieldValue($str);
+        $headers->addHeader($header);
+        $message->setHeaders($headers);
+        $headers->setEncoding('UTF-8');
+        
+		$this->assertTrue(is_string($header->toString()), 'Can call Header::toString()');
```
Comment (Maks3w) [inline test/HeadersTest.php]: 1. assertTrue(is_Foo) can be replaced with assertInternalType(foo)
2. If test is about an encoding problem is better to test output value is exactly the expected encoded value.

File: src/Headers.php [hunk 5]
```diff
@@ -152,7 +152,7 @@ public function setEncoding($encoding)
     {
```
Comment (Maks3w) [inline src/Headers.php]: I started to think all this function is a bad solution for proper header encoding. Header encoding should be done by each particular case.

Could we deprecate/remove this function? /cc @weierophinney

File: test/HeadersTest.php (hunk 4 above)
Comment (leonardo-piseri) [inline test/HeadersTest.php]: The problem is not an incorrect encoding, but just the ability to add non ascii headers to a mail message.  The goal of the test is to show that If you create a Zend\Mail\Header() with non ascii value, you can't call toString Method after you add it to a Zend\Mail\Message().
I really don't have good experience in writing tests, I know!

**Comment by weierophinney** [comment]
Per @Maks3w, I'm unsure the test is actually testing the behavior you intend. One way to clarify it is to write it out in plain language first, and then try and write a test to exercise the behavior:

> - When I ...
> - Then I expect ...

Right now, looking at the test, I'm unsure what behavior is being tested; I'd assume that `Header::toString()` would return a string already; what might be more interesting is asserting against the contents of the string — and/or commenting that prior to the fix, an exception occurred at that time. (Assertions against the string contents would still make sense, however!)

**Comment by Maks3w** [comment]
Closed due inactivity

**Pull request closed without merging**